Improve performance of AllBeEquivalentTo

krajek opened the pull request:
Attempt to fix #914.

There are (at least) two bottlenecks:

- [x] `AllBeEquivalentTo` created a collection of items matching the subject's length, and then used loose ordering which results in O(n^2) complexity. Change to strict ordering is necessary to have hope for any kind of reasonable performance.
- [x] Building collection of messages. The more inequivalent items there is the more time it takes to create assert failure messages. I think that some kind of early break is needed in `EnumerableEquivalencyValidator` after say 10 failures.
- [x] Extract `OrderingRules.IsOrderingStrictFor(context)` in `EnumerableEquivalencyValidator.AssertElementGraphEquivalency` before the loop. During profiling, it stands out as an easy-fix with the execution time around 8-10% spent in this method.

I checked that after those three issues there are no more outstanding hot spots. Execution time is split among some `CanHandle` methods, creating objects like `EquivalencyValidationContext` and so on.

Any ideas are warmly welcomed.

jnyrup commented on Tests/Shared.Specs/CollectionEquivalencySpecs.cs:
```diff
@@ -900,14 +900,45 @@ public void When_all_subject_items_are_not_equivalent_to_expectation_object_it_s
+            //-----------------------------------------------------------------------------------------------------------
+            // Arrange
+            //-----------------------------------------------------------------------------------------------------------
+            const int N = 100000;
```
🔧 How about using `Enumerable.Range(1, 100_000).ToList()`?

krajek replied:
Notice that first item is not equal to the rest. If the test will ultimately use single value I will refactor to `Enumerable.Range`.

jnyrup commented on Tests/Shared.Specs/CollectionEquivalencySpecs.cs:
```diff
@@ -900,14 +900,45 @@ public void When_all_subject_items_are_not_equivalent_to_expectation_object_it_s
+                }
+            };
+
+            ////-----------------------------------------------------------------------------------------------------------
```
🙃 `////`

jnyrup commented:
Out of curiosity, Did you measure the wall time improvement of this PR for the new test?

krajek commented:
What does "wall time" mean? :-)

jnyrup commented:
Wall time is just another word for the runtime measured in e.g. seconds instead of algorithmic complexity.
E.g. if applying this PR improves the test from 100s to 10s.

krajek commented:
I tried to measure the time for 10000 items but gave up after a few minutes :-). Even for a few hundreds of items times were in the order of minutes. With strict ordering and 100 000 items, I have around 13s.

krajek commented:
FYI: Last commit(502977c) fixed issue unrelated to the topic of the PR. When I set strict ordering for the top-level collection exception message for existing test case changed. The reason was the discrepancy between `LooselyMatchAgainst` and `StrictlyMatchAgainst`.
`LooselyMatchAgainst` created scope for each item whereas `StrictlyMatchAgainst` didn't. 

Please review that particular scenario, I am not sure I understand the code deeply enough.

dennisdoomen commented:
I'm loosing track of what each PR is supposed to. Normally, if a PR consists of multiple commits, I assume you have carefully squashed individual changes into a couple of well-focused commits. We also merge these PRs by keeping all commits. So please either separate unrelated changes into separate PRs or make sure you squash commits that change stuff in previous commits into the same one. I regularly use [fixup](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html) commits and/or amend existing commits.

dennisdoomen commented on Src/FluentAssertions/Equivalency/EnumerableEquivalencyValidator.cs:
```diff
@@ -62,35 +64,73 @@ private void AssertElementGraphEquivalency<T>(object[] subjects, T[] expectation
+                        if (failedCount >= FailedItemsFastFailThreshold)
+                        {
+                            context.TraceSingle(path =>
+                                $"Fail failing strict order comparison of collection after {FailedItemsFastFailThreshold} items failed at {path}");
```
🤔 I _think_ you mean that you already found 10 mismatches and decided to abort the comparison for performance reasons, right?

krajek replied:
"Fail failing strict order comparison" - jeez what was I thinking? Fixed.

dennisdoomen commented on Src/FluentAssertions/Execution/AssertionScope.cs:
```diff
@@ -284,6 +284,12 @@ public string[] Discard()
             return assertionStrategy.DiscardFailures().ToArray();
         }
 
+        public bool AnyFailures()
```
🤔 Why is it a method and not a property? 
🙃 I would call it `HasFailures`

dennisdoomen commented on Tests/Shared.Specs/CollectionEquivalencySpecs.cs:
```diff
@@ -900,14 +922,66 @@ public void When_all_subject_items_are_not_equivalent_to_expectation_object_it_s
+                try { subject.Should().AllBeEquivalentTo(1); }
+                catch (Exception)
+                {
+                    // ignored, we only care about execution time
```
👍 That _is_ a very helpful comment.

dennisdoomen commented on Src/FluentAssertions/Equivalency/EnumerableEquivalencyValidator.cs:
```diff
@@ -137,9 +179,20 @@ private string[] TryToMatch<T>(object subject, T expectation, int expectationInd
+            {
+                object subject = subjects[expectationIndex];
+                string indexString = expectationIndex.ToString();
+                var equivalencyValidationContext = context
```
🙃 In this code-base, we prefer explicit types, unless the type itself is clear already.

krajek replied:
Ok, fixed.

dennisdoomen commented on Src/FluentAssertions/Equivalency/EnumerableEquivalencyValidator.cs:
```diff
@@ -62,35 +64,73 @@ private void AssertElementGraphEquivalency<T>(object[] subjects, T[] expectation
+                    $"Finding the best match of {expectation} within all items in {subjects} at {path}[{index}]"))
                 {
-                    using (context.TraceBlock(path => $"Strictly comparing expectation {expectation} at {path} to item with index {index} in {subjects}"))
+                    bool succeed = LooselyMatchAgainst(subjects, expectation, index);
```
🙃 It's either `success` or `succeeded`

krajek commented on Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:
```diff
@@ -331,12 +331,22 @@ private bool IsValidProperty<TSelector>(Expression<Func<T, TSelector>> propertyE
 
-            BeEquivalentTo(repeatedExpectation, config, because, becauseArgs);
+            // Because we have just manually created the collection based on single element
+            // we are sure that we can force string ordering, because ordering does not matter in terms
```
Good catch

krajek commented on Tests/Shared.Specs/CollectionEquivalencySpecs.cs:
```diff
@@ -324,6 +324,28 @@ public void When_a_collection_does_not_match_it_should_include_items_in_message(
+            // Arrange
+            //-----------------------------------------------------------------------------------------------------------
+            const int commonLength = 11;
+            var subject = new int[commonLength] { 0, 1, 2, 2, 4, 5, 6, 7, 8, 9, 10 };
```
Repeated 2 fixed, I have added a comment explaining why I'm not using `Enumerable.Repeat` this time

krajek commented:
First of all thanks for mentioning fixup commits, I had no idea such a thing exists.
Secondly, thanks for describing the merging policy. I am more used to no-constraints feature branches merged with squash to the main branch. I would be fine with such final for this PR as well.

Regarding the topic of this PR. Every commit but one is directly tackling performance problem as stated in the PR title. The one outstanding commit (502977c) was necessary (at least at the time of writing) because after I introduced the `WithStrictOrdering` usage I accidentally reproduced the #918 issue. The fix in my case what to introduce new `AssertionScope` for strict ordering path just like a loose ordering path was already doing.

Now we need to decide how to proceed. Possible scenarios:

A) After my explanation, you understand the scope of the feature and you merge it with squash.
B) I abandon this particular PR and split it into three, one for each "sub-task".
@dennisdoomen would you be ok with A) this time?

In either case from now on, for new PR's I will prepare my branches
1) to be as small as possible
2) to be prepared for a merge with all the commits using fixup/amend/rebase etc.

dennisdoomen commented:
> would you be ok with A) this time?

Absolutely. My intention was just to explain my preference for this repo.
